### ISSUE TITLE
More than one initializer found bug

### ISSUE THREAD

**Issue by NathanaelA** [NativeScript/ios-runtime#259]
Using v1.20 runtimes & v1.20 common, and v1.2.1 tns command.

Ok, this is a REALLY weird issue, and I've done as much debugging as I can at this point.   I've been working on a cool websocket library and now have it fully working on IOS and Android properly.  So I am trying to make it into a easy to use plugin and do some final tests; so I change
My applications require statement to
**require('nativescript-websockets');**
and the nativescript-websockets.js require to
**require('nativescript-websockets/websocket-common');**
and then install the plugin using the **tns plugin add nativescript-websockets**  

And everything installs properly and I see my directory and plugin in the platforms/ios/app/tns_modules folder like I expect.    

Now If I try to do **tns run ios --emulator** I get the following error message:
```
/app/tns_modules/file-system/file-system.js:314:24: JS ERROR Error: More than one initializer found that matches constructor invocation: initWithCapacity: initWithArray: initWithArray:copyItems: init
1   0xe7c30 NativeScript::FFICallback<NativeScript::ObjCMethodCallback>::ffiClosureCallback(ffi_cif*, void*, void**, void*)
2   0x6b1796 ffi_closure_SYSV_inner
3   0x6b11e2 .LCFI7
4   0x1710f24 -[UIViewController loadViewIfRequired]
5   0x1711215 -[UIViewController view]
6   0x1608215 -[UIWindow addRootViewControllerViewIfPossible]
7   0x1608689 -[UIWindow _setHidden:forced:]
8   0x1608940 -[UIWindow _orderFrontWithoutMakingKey]
9   0x1616c7d -[UIWindow makeKeyAndVisible]
10  0x6b10fd .LCFI1
11  0x6b16df ffi_call
12  0xe6511 NativeScript::ObjCMethodCall::derivedExecuteCall(JSC::ExecState*, unsigned char*)
13  0xe6808 long long NativeScript::FFICall::executeCall<NativeScript::ObjCMethodCall>(JSC::ExecState*)
14  0x45fc1e JSC::LLInt::setUpCall(JSC::ExecState*, JSC::Instruction*, JSC::CodeSpecializationKind, JSC::JSValue, JSC::LLIntCallLinkInfo*)
15  0x45c056 llint_slow_path_call
16  0x46497f llint_entry
17  0x4604c6 vmEntryToJavaScript
18  0x3f637c JSC::JITCode::execute(JSC::VM*, JSC::ProtoCallFrame*)
19  0x3d7cc5 JSC::Interpreter::executeCall(JSC::ExecState*, JSC::JSObject*, JSC::CallType, JSC::CallData const&, JSC::JSValue, JSC::ArgList const&)
20  0x53bd8d JSC::call(JSC::ExecState*, JSC::JSValue, JSC::CallType, JSC::CallData const&, JSC::JSValue, JSC::ArgList const&)
21  0xe7b1b NativeScript::FFICallback<NativeScript::ObjCMethodCallback>::callFunction(JSC::JSValue const&, JSC::ArgList const&, void*)
22  0xe6eae NativeScript::ObjCMethodCallback::ffiClosureCallback(void*, void**, void*)
23  0xe7bf4 NativeScript::FFICallback<NativeScript::ObjCMethodCallback>::ffiClosureCallback(ffi_cif*, void*, void**, void*)
24  0x6b1796 ffi_closure_SYSV_inner
25  0x6b11e2 .LCFI7
26  0x15b28d0 -[UIApplication _handleDelegateCallbacksWithOptions:isSuspended:restoreState:]
27  0x15b365b -[UIApplication _callInitializationDelegatesForMainScene:transitionContext:]
28  0x15b6cb2 -[UIApplication _runWithMainScene:transitionContext:completion:]
29  0x15cfe1e __84-[UIApplication _handleApplicationActivationWithScene:transitionContext:completion:]_block_invoke
30  0x15b58da -[UIApplication workspaceDidEndTransaction:]
31  0x3aa890e __37-[FBSWorkspace clientEndTransaction:]_block_invoke_2
```

Now, If I **COPY** (or move) the two files from the platforms/ios/tns_modules/nativescript-websockets back in to the main app folder; and just use require('./nativescript-websockets') and change the nativescript-websocket.js to require('./websockets-common') everything **works** as expected.    

I have tried as debugging steps:
1. Renaming all the files; and verifying the package.json matches.  (same issue occurs)
2. Renaming just the websockets-common file; which then the emulator will actually throw a error about the require unable to find it -- which is proper since that file is now "missing" under its name that the nativescript-websockets file is trying to require.   So we know that the files are being found.
3. If I remove the require from the nativescript-websockets; and then include EITHER file by itself; it works.   If I include both, it crashes.

Here is the app.js
```js
var application = require("application");
application.mainModule = "main-page";
application.cssFile = "./app.css";
// require('nativescript-websockets');
// require('nativescript-websockets/websockets-common.js');
application.start();
```

So I can uncomment either line; but not both.   If I uncomment both; the app will crash with the exact same above error.    Just like if I let the nativescript-websocket require the websocket-common file.

 And none of my websocket code actually does a require of the "file-system"; I don't actually require anything from the nativescript common core.

So in summary; if it is included from where the plugin stuck it in the platforms/ios/app/tns_modules it will crash.   If I manually copy both these files into my app folder and fix the app require and the nativescript-websockets require to both point to the app copy; the app works.

Gotta love this bug....

**Comment by KristinaKoeva** [comment]
Hey Nathanael,

We are aware of this issue and we are going to provide a fix for it in the next few days. 
The exception is somehow misleading because it doesn't come directly from the file-system module. The code that fails is located in tns_modules/utils/utils.ios.js and the error comes from this method  
```
function jsArrayToNSArray(str) {
            var arr = new NSMutableArray();
            if ("undefined" !== typeof str) {
                for (var element in str) {
                    arr.addObject(str[element]);
                }
            }
            return arr;
}
```

As a workaround you could change the implementation of the jsArrayToNSArray to 
```
function jsArrayToNSArray(str) {
            return str;
}
``` 
and since we have integrated adapters for NSArray (https://github.com/NativeScript/ios-runtime/pull/64) it would work as expected.

**Comment by NathanaelA** [comment]
@KristinaKoeva awesome -- I was worried that this was going to be a issue to debug since it made no sense.  
I only have a couple small concerns:
1. why does loading a module from the platform/ios/tns_modules folder vs the root apps folder causes the issue.   Is their another issue in the runtime that runs a different code path?  
2. Do we know why the error is propagated from a different location.

**Comment by KristinaKoeva** [comment]
Hey Nathanael,

To your questions
1. I am not exactly sure why the application crashes in the first case but not the other.
    Here is the call stack 
```
NSMutableArray@[native code]
jsArrayToNSArray@/app/tns_modules/utils/utils.js:26:41
concatPath@/app/tns_modules/file-system/file-system-access.js:131:67
enumEntities@/app/tns_modules/file-system/file-system-access.js:275:42
eachEntity@/app/tns_modules/file-system/file-system-access.js:97:26
eachEntity@/app/tns_modules/file-system/file-system.js:311:35
getFileCandidatesFromFolder@/app/tns_modules/file-system/file-name-resolver.js:113:30
resolveFileNameImpl@/app/tns_modules/file-system/file-name-resolver.js:103:58
resolveFileName@/app/tns_modules/file-system/file-name-resolver.js:94:46
resolveFileName@/app/tns_modules/file-system/file-name-resolver.js:190:44
pageFromBuilder@/app/tns_modules/ui/frame/frame-common.js:66:54
resolvePageFromEntry@/app/tns_modules/ui/frame/frame-common.js:54:35
navigate@/app/tns_modules/ui/frame/frame-common.js:112:40
navigate@/app/tns_modules/ui/frame/frame.js:27:43
onLoaded@/app/tns_modules/ui/frame/frame.js:21:26
viewDidLoad@/app/tns_modules/ui/frame/frame.js:157:29
makeKeyAndVisible@[native code]
applicationDidFinishLaunchingWithOptions@/app/tns_modules/application/application.js:64:38
UIApplicationMain@[native code]
start@/app/tns_modules/application/application.js:148:26
anonymous@/app/app.js:4:18
__executeModule
__loadModule
```

2. The original error is rethrown so the original location is not preserved
```
var onError = function (error) {
    throw error;
};
getFileAccess().eachEntity(this.path, onSuccess, onError);
```

ping @hshristov to shed some light on both questions

**Comment by NathanaelA** [comment]
I hope v1.2.2 gets released soon.  I just pushed my nativescript-websockets.   And created an issue to show people how to work around it.  But it would be much nicer if they don't have to.

**Comment by fealebenpae** [comment]
Hey @NathanaelA , v1.2.2 has been our for a week, do you still experience the "more than one initializer" problem on a fresh app without @KristinaKoeva's workaround?

**Comment by NathanaelA** [comment]
@fealebenpae - Yes, I just did a fully fresh application and it pulled v1.2.2 and v1.2.2 works as expected.

**Comment by vchimev** [comment]
I experienced the same issue while testing the forthcoming 1.3.0 version. Ping me for a repro project if needed.

**Comment by fealebenpae** [comment]
@vchimev if you still have a project that reproduces this, it would be more than helpful, thanks.

**Comment by vchimev** [comment]
@fealebenpae, my project works smoothly and does not reproduce the issue any more.

**Comment by fealebenpae** [comment]
Great, I'll close this, then.
